website/pages/blog/building-cloudquery.md

```diff
@@ -0,0 +1,102 @@
+
+# Technical Deep Dive
+
+This section will discuss the most important decisions and solutions we landed on, organized by the challenges listed above.
+
+## API Coverage
+
+### Open Source
+
+There is no vendor that can support all integrations in-house thus the best and only way in our opinion is to go the open-source way and give the ability for others to contribute to our official plugins and develop community plugins without being blocked by a single vendor. Good recent examples that dealt with the “infinite APIs integrations” problem and took the open source route successfully are Terraform and Pulumi.
+
+### Pluggable Sources
+
+Pluggable sources are key to scale API coverage both from development and usability perspective and solve the following challenges:
+
+- **Binary Size:** Users download only the plugins they want to use to ensure good performance and user experience without a binary that grows infinitely and will eventually cause not only binary size issues but library clashes from different services.
+- **Versioning:** Given plugins are developed as standalone gRPC binaries users can use different versions of different plugins depending on their needs.
+- **Independent development:** Developers can develop their own plugins in their own repositories without being blocked by a vendor (us in this case) to review PRs.
+
+Our [pluggable system](https://www.cloudquery.io/docs/developers/architecture) is based on gRPC to ensure our plugins can be cross-platform, independent  and performant.
+
+### SDK for Source Plugins
+
+Writing ELT code involves a lot of boilerplate code, testing code, tricky performance and parallelization code. We’ve built a [Go SDK](https://github.com/cloudquery/plugin-sdk/) that enables both us and other developers to focus only on the E(Extract) code while CloudQuery SDK takes care of parallelization, transformation, loading and testing.
+
+### Code Generation
+
+Big challenge is scaling and continuously supportting large number of APIs. Before we dive into how we leveraged code-generation to generate CloudQuery plugins, let’s take a quick look at how client library generation for multiple languages worked in those days.
+
+Usually there is some intermediate language such as gRPC/GraphQL/Smithy that generates server stubs and clients automatically for multiple languages instead of manually maintaining clients for a number of languages and keeping up with the server APIs.
+
+There were some significant advances around client library generation in the recent years that we also took advantage of.
+
+We could take advantage of the schema definition and generate the source plugin from that, but we decided to do that one step further in the pipeline from the Go Code.
+
+Go is incredibly suited for Code Analysis so we created a [code generation library](https://github.com/cloudquery/plugin-sdk/tree/main/codegen) that generates CloudQuery schema from Go structs. This helps achieve number of things:
```

Comment:
```suggestion
Go is incredibly suited for Code Analysis so we created a [code generation library](https://github.com/cloudquery/plugin-sdk/tree/main/codegen) that generates CloudQuery schema from Go structs. This helps achieve a number of things:
```

Reply:
`a number of things`? I wont argue but seems odd to me as it's more than one ?

Reply:
Definitely `a number of things` - I guess because the article is referring to `number`, which is one thing, but English can be funny 🤷

Reply:
done